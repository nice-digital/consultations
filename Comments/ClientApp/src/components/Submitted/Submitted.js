// @flow

import React, { Component } from "react";
import { withRouter } from "react-router-dom";
import { tagManager } from "../../helpers/tag-manager";
import Helmet from "react-helmet";
import BreadCrumbsWithRouter from "../Breadcrumbs/Breadcrumbs";
import { Header } from "../Header/Header";
import { UserContext } from "../../context/UserContext";
import LoginBannerWithRouter from "../LoginBanner/LoginBanner";
import preload from "../../data/pre-loader";
import { load } from "../../data/loader";
import { SubmittedContent } from "../SubmittedContent/SubmittedContent";

type PropsType = any;

type StateType = {
	consultationData: ConsultationStateType | null,
	loading: boolean,
	hasInitialData: boolean,
	error: {
		hasError: boolean,
		message: string,
	},

}

export class Submitted extends Component<PropsType, StateType> {

	constructor(props: PropsType) {
		super(props);

		this.state = {
			consultationData: null,
			loading: true,
			hasInitialData: false,
			error: {
				hasError: false,
				message: "",
			},
		};

		if (this.props) {

			let preloadedConsultation;

			let preloadedData = {};
			if (this.props.staticContext && this.props.staticContext.preload) {
				preloadedData = this.props.staticContext.preload.data; //this is data from Configure => SupplyData in Startup.cs. the main thing it contains for this call is the cookie for the current user.
			}

			preloadedConsultation = preload(
				this.props.staticContext,
				"consultation",
				[],
				{consultationId: this.props.match.params.consultationId, isReview: false},
				preloadedData,
			);

			if (preloadedConsultation) {
				if (this.props.staticContext) {
					this.props.staticContext.analyticsGlobals.gidReference = preloadedConsultation.reference;
					this.props.staticContext.analyticsGlobals.consultationTitle = preloadedConsultation.title;
					this.props.staticContext.analyticsGlobals.stage = "submit";
				}
				this.state = {
					consultationData: preloadedConsultation,
					loading: false,
					hasInitialData: true,
					error: {
						hasError: false,
						message: "",
					},
				};
			}
		}
	}

	gatherData = async () => {
		const {consultationId} = this.props.match.params;

		const consultationData = load("consultation", undefined, [], {
			consultationId,
			isReview: false,
		})
			.then(response => response.data)
			.catch(err => {
				console.log(err);
			});

		return {
			consultationData: await consultationData,
		};
	};

	componentDidMount() {
		if (!this.state.hasInitialData) {
			this.gatherData()
				.then(data => {
					this.setState({
						consultationData: data.consultationData,
						loading: false,
						hasInitialData: true,
					}, () => {
						tagManager({
							event: "pageview",
							gidReference: this.state.consultationData.reference,
							title: this.getPageTitle(true),
							stage: "submit",
						});
					});
				})
				.catch(err => {
					this.setState({
						error: {
							hasError: true,
							message: "gatherData in componentDidMount failed " + err,
						},
					});
				});
		}
	}

	getPageTitle = (isForAnalytics: boolean = false) => {
		if (isForAnalytics) return this.state.consultationData.title;
		return `${this.state.consultationData.title} | Response submitted`;
	};

	render() {
		if (!this.state.hasInitialData) return <h1>Loading...</h1>;
		return (
			<>
				<Helmet>
					<title>{this.getPageTitle()}</title>
				</Helmet>
				<div className="container">
					<div className="grid">
						<div data-g="12">
							<BreadCrumbsWithRouter links={this.state.consultationData.breadcrumbs}/>
							<main>
								<div className="page-header">
									<Header
										title="Response submitted"
										consultationState={this.state.consultationData.consultationState}
									/>
									<UserContext.Consumer>
										{(contextValue: ContextType) => {
											return (
												!contextValue.isAuthorised ?
													<LoginBannerWithRouter
														signInButton={true}
														currentURL={this.props.match.url}
														signInURL={contextValue.signInURL}
														registerURL={contextValue.registerURL}
														allowOrganisationCodeLogin={false}
														orgFieldName="submitted"
													/> :
													<>
														<SubmittedContent
															organisationName={contextValue.organisationName}
															isOrganisationCommenter={contextValue.isOrganisationCommenter}
															isLead={contextValue.isLead}
															consultationState={this.state.consultationData.consultationState}
															consultationId={this.props.match.params.consultationId}
															basename={this.props.basename}
															isSubmitted={true}
															linkToReviewPage={true}
														/>

														<h2>Help us improve our online commenting service</h2>
														<p>This is the first time we have used our new online commenting software. We'd really like to hear your feedback so that we can keep improving
															it.</p>
														<p>Answer our short, anonymous survey (4 questions, 2 minutes).</p>
														<p>
															<a className="btn btn--cta"
																href="https://surveys.hotjar.com/62df646f-8edb-4c03-8275-3ec8a75d0214"
																target="_blank"
																rel="noopener noreferrer">
																Answer the survey
															</a>
														</p>
													</>
											);
										}}
									</UserContext.Consumer>
								</div>
							</ main>
						</div>
					</div>
				</div>
			</>

		);
	}
}

export default withRouter(Submitted);
