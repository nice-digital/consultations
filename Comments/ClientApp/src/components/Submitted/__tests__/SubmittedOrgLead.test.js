import React from "react";
import { render, waitForElementToBeRemoved, screen } from "@testing-library/react";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import { MemoryRouter } from "react-router";
import { Submitted } from "../Submitted";
import ConsultationData from "./Consultation.json";

jest.mock("../../../context/UserContext", () => {
	return {
		UserContext: {
			Consumer: (props) => {
				return props.children({
					isAuthorised: true,
					isOrganisationCommenter: true,
					isLead: true,
					organisationName: "Really Cool Org",
				});
			},
		},
	};
});

test("should match snapshot with supplied data", () => {
	const mock = new MockAdapter(axios);
	let consultationPromise = new Promise(resolve => {
		mock
			.onAny()
			.reply(() => {
				resolve();
				return [200, ConsultationData];
			});
	});
	const fakeProps = {
		match: {
			params: {
				consultationId: 1,
			},
		},
	};
	const {container} = render(<Submitted {...fakeProps} />, {wrapper: MemoryRouter});
	return Promise.all([consultationPromise]).then(async () => {
		await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "h1" }));
		expect(container).toMatchSnapshot();
	});
});
