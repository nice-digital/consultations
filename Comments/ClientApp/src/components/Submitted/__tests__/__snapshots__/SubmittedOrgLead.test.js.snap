// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot with supplied data 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="grid"
    >
      <div
        data-g="12"
      >
        <nav
          aria-label="Breadcrumbs"
        >
          <p
            class="visually-hidden"
          >
            You are here:
          </p>
          <ol
            class="breadcrumbs"
            itemscope=""
            itemtype="http://schema.org/BreadcrumbList"
          >
            <li
              class="breadcrumbs__crumb"
              itemprop="itemListElement"
              itemscope=""
              itemtype="http://schema.org/ListItem"
              role="presentation"
            >
              <a
                href="/"
                itemprop="item"
              >
                <span
                  itemprop="name"
                >
                  Home
                </span>
              </a>
              <meta
                content="0"
                itemprop="position"
              />
            </li>
            <li
              class="breadcrumbs__crumb"
              itemprop="itemListElement"
              itemscope=""
              itemtype="http://schema.org/ListItem"
              role="presentation"
            >
              <a
                href="/guidance/indevelopment/gid-ta10232/consultation/"
                itemprop="item"
              >
                <span
                  itemprop="name"
                >
                  Consultation title
                </span>
              </a>
              <meta
                content="1"
                itemprop="position"
              />
            </li>
            <li
              class="breadcrumbs__crumb"
              itemprop="itemListElement"
              itemscope=""
              itemtype="http://schema.org/ListItem"
              role="presentation"
            >
              <a
                href="/consultations/1/1/introduction"
                itemprop="item"
              >
                <span
                  itemprop="name"
                >
                  Consultation documents
                </span>
              </a>
              <meta
                content="2"
                itemprop="position"
              />
            </li>
          </ol>
        </nav>
        <main>
          <div
            class="page-header"
          >
            <h1
              class="page-header__heading mt--0"
              data-qa-sel="changeable-page-header"
            >
              Response submitted
            </h1>
            <div
              class="mb--d"
            >
              <p
                class="container container-full ml--0"
              >
                <span>
                  <span
                    class="tag tag--open"
                  >
                    Open for comments
                  </span>
                   Open until
                   
                  <time
                    datetime="1874937600000"
                  >
                    31 May 2029
                  </time>
                </span>
                  
                <a
                  href="/leadinformation"
                >
                  Request commenting lead permission
                </a>
              </p>
            </div>
            <h2>
              Your response was submitted 
               on 
              <time
                datetime="1563197058473"
              >
                15 July 2019
              </time>
              .
            </h2>
            <ul
              class="list list--unstyled"
            >
              <li>
                <a
                  data-qa-sel="review-submitted-comments"
                  href="/1/review"
                >
                  Review your response
                </a>
              </li>
              <li>
                <a
                  href="undefined/api/exportexternal/1"
                >
                  Download submitted response (Excel)
                </a>
              </li>
              <li>
                <a
                  href="undefined/api/exportlead/1"
                >
                  Download all responses from your organisation (Excel)
                </a>
              </li>
            </ul>
            <h2>
              What happens next?
            </h2>
            <p>
              We will review all the submissions received for this consultation. Our response will be published on the website around the time the guidance is published.
            </p>
            <hr />
            <h2>
              Help us improve our online commenting service
            </h2>
            <p>
              This is the first time we have used our new online commenting software. We'd really like to hear your feedback so that we can keep improving it.
            </p>
            <p>
              Answer our short, anonymous survey (4 questions, 2 minutes).
            </p>
            <p>
              <a
                class="btn btn--cta"
                href="https://surveys.hotjar.com/62df646f-8edb-4c03-8275-3ec8a75d0214"
                rel="noopener noreferrer"
                target="_blank"
              >
                Answer the survey
              </a>
            </p>
          </div>
        </main>
      </div>
    </div>
  </div>
</div>
`;
